### DISCUSSION TITLE
ContinuationToken with Issues

### DISCUSSION THREAD

**Discussion by DuPlooydeLange** [iTwin/itwinjs-core#4586] [Q&A] [answered]
I am trying to wrap my head around the expectation of the continuation token and paginated results when working with issues. I am using the API to call data to PowerBI and I am only getting 50 results 

There is a great [blog post](https://medium.com/itwinjs/itwin-clash-detection-api-how-to-46339457c79a) about get the data but it's severely limited

Any help would be greatly appreciated

**Comment by Johannes-repos** [comment, marked as answer]
Hello,
Once you get the results into PowerBI you will see a column called _links.next.href. This contains the continuation token which can be used for a new query to get the next 50 results.
Use something like:
https://api.bentley.com/issues/?continuationToken=eyJ0b3AiOjEwMDA....&projectId=" & ProjectId, [Headers=[Prefer="minimal", Accept="application/vnd.bentley.itwin-platform.v1+json", Authorization="Bearer " & #"Token"]])),

You can test this with the _Try it out_ feature.
https://developer.bentley.com/apis/issues-v1/operations/get-project-issues/
Run the query first with only the projectId and authorization. Once the result appears, scroll down to the end of the list, and look out for the continuation token. Copy and paste the token into the parameter field and trigger the query again.
HTH

**Reply by DuPlooydeLange** [reply]
Thanks @Johannes-repos

**Reply by DuPlooydeLange** [reply]
@Johannes-repos .

Thanks again for your support but I am really struggling with the continuation token in Power Query, maybe you have some advise?

Here is what I came up with, any advise would greatly appreciated.

` 

let
           
    // Create a function to build the uri for the API
    BuildUri = (Projectid as text, continuationToken as nullable text) => 
    if continuationToken = null then
        "https://api.bentley.com/issues/?$top=1000&projectId=" & Projectid
    else
        "https://api.bentley.com/issues/?projectId=" & Projectid & "&continuationToken=" & continuationToken,
        
    // Define the API Call
    InitialIssues = Json.Document(
        Web.Contents(
            BuildUri(Projectid, null),
            [
                Headers = [
                    Prefer = "return=representation", 
                    Accept = "application/vnd.bentley.itwin-platform.v1+json", 
                    Authorization = "Bearer " & #"GETAccessToken"()
                ]
            ]
        )
    ), 
    
    // Get the continuation token from the response
    ContinuationToken = if Uri.Parts(InitialIssues[_links][next][href]) = null then null else Uri.Parts(InitialIssues[_links][next][href])[Query][continuationToken], 
    FirstIssues = InitialIssues[issues], 

    // Create a function to retrieve more data from the API using the continuation token
    GetMoreData = (ContinuationToken as text, Issues as list) =>
    try 
        let
            GetIssues = Json.Document(
                Web.Contents(
                    BuildUri(Projectid, ContinuationToken), 
                    [
                        Headers = [
                            Prefer = "return=representation", 
                            Accept = "application/vnd.bentley.itwin-platform.v1+json", 
                            Authorization = "Bearer " & #"GETAccessToken"()
                        ]
                    ]
                )
            ), 
            Issues = GetIssues[issues],
            NextContinuationToken = if Uri.Parts(GetIssues[_links][next][href]) = null then null else Uri.Parts(GetIssues[_links][next][href])[Query][continuationToken]
        in
        [NextContinuationToken = NextContinuationToken, Issues = Issues],

    // Recursive function to call GetMoreData until the continuation token is null
    RecursiveGetIssues = (ContinuationToken as text, AccumulatedIssues as list) =>
    let
        NextData = GetMoreData(ContinuationToken),
        NextContinuationToken = NextData[Value][NextContinuationToken],
        NextIssues = NextData[Value][Issues]
    in
        if NextContinuationToken = null then
            AccumulatedIssues & NextIssues
        else
            RecursiveGetIssues(NextContinuationToken, AccumulatedIssues & NextIssues)
in
    RecursiveGetIssues(ContinuationToken = ContinuationToken, FirstIssues = FirstIssues)

 `

**Reply by Johannes-repos** [reply]
Hello again,

we created a simple sample PowerBI file for you. https://github.com/Johannes-repos/IssuesAPI

You can use this to load all project issues in one go. Simply replace the required parameters and you're good to go. Please report back how it is working for you.

Out of interest; what kind of applications are you planning to create using the iTwin Platform? We got a number of samples at developer.bentley.com

**Reply by DuPlooydeLange** [reply]
Hi @Johannes-repos

I cannot thank you enough this works perfect.
Its been months of frustration to get all our project issues loaded into PowerBI.

Really appreciate your time and assistance in this matter.

Dup
